// gatsby-cli/lib/reporter/redux/index.js
"use strict";

exports.__esModule = true;
exports.setStore = exports.onLogAction = exports.onStoreSwap = exports.dispatch = exports.getStore = void 0;

var _redux = require("redux");

var _reducer = require("./reducer");

var _utils = require("./utils");

var _constants = require("../constants");

function _createForOfIteratorHelperLoose(o) { var i = 0; if (typeof Symbol === "undefined" || o[Symbol.iterator] == null) { if (Array.isArray(o) || (o = _unsupportedIterableToArray(o))) return function () { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }; throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } i = o[Symbol.iterator](); return i.next.bind(i); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }

let store = (0, _redux.createStore)((0, _redux.combineReducers)({
  logs: _reducer.reducer
}), {});
const storeSwapListeners = [];
const onLogActionListeners = new Set();

const getStore = () => store;

exports.getStore = getStore;

const dispatch = action => {
  if (!action) {
    return;
  }

  if (Array.isArray(action)) {
    action.forEach(item => dispatch(item));
    return;
  } else if (typeof action === `function`) {
    action(dispatch);
    return;
  }

  action = Object.assign({}, action, {
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore this is a typescript no-no..
    // And i'm pretty sure this timestamp isn't used anywhere.
    // but for now, the structured logs integration tests expect it
    // so it's easier to leave it and then explore as a follow up
    timestamp: new Date().toJSON()
  });
  store.dispatch(action);

  if ((0, _utils.isInternalAction)(action)) {
    // consumers (ipc, yurnalist, json logger) shouldn't have to
    // deal with actions needed just for internal tracking of status
    return;
  }

  for (var _iterator = _createForOfIteratorHelperLoose(onLogActionListeners), _step; !(_step = _iterator()).done;) {
    const fn = _step.value;
    fn(action);
  }
};

exports.dispatch = dispatch;

const onStoreSwap = fn => {
  storeSwapListeners.push(fn);
};

exports.onStoreSwap = onStoreSwap;

const onLogAction = fn => {
  onLogActionListeners.add(fn);
  return () => {
    onLogActionListeners.delete(fn);
  };
};

exports.onLogAction = onLogAction;

const setStore = s => {
  s.dispatch({
    type: _constants.Actions.SetLogs,
    payload: store.getState().logs
  });
  store = s;
  storeSwapListeners.forEach(fn => fn(store));
};

exports.setStore = setStore;
